feat: Allow indirect replication

patriknw opened the pull request:
Draft because requires a change in Akka first https://github.com/akka/akka/pull/32233

patriknw commented on akka-projection-grpc-tests/src/it/scala/akka/projection/grpc/replication/IndirectReplicationIntegrationSpec.scala:
```diff
@@ -0,0 +1,297 @@
+import org.scalatest.wordspec.AnyWordSpecLike
+import org.slf4j.LoggerFactory
+
+object IndirectReplicationIntegrationSpec {
```
This is similar to existing ReplicationIntegrationSpec

patriknw commented on akka-projection-grpc-tests/src/test/scala/akka/projection/grpc/replication/ReplicationSettingsSpec.scala:
```diff
@@ -32,24 +32,36 @@ class ReplicationSettingsSpec extends AnyWordSpec with Matchers {
+           # only events from the origin replica will be transferred from the origin
+           # replica. When each replica is connected to each other replica it's most
+           # efficient to disable indirect replication.
+           indirect-replication = on
```
default value is `off`, this is included in docs and maybe we should show `off`
however, then the test would be testing this

patriknw replied:
yes

patriknw commented on akka-projection-grpc/src/main/scala/akka/projection/grpc/replication/internal/ReplicationImpl.scala:
```diff
@@ -86,8 +86,10 @@ private[akka] object ReplicationImpl {
+          if (settings.indirectReplication || meta.originReplica == settings.selfReplicaId)
+            Future.successful(envelope.eventOption)
+          else
+            filteredEvent // Optimization: was replicated to this replica, don't pass the payload across the wire
```
This is the reason why I haven't enabled it by default. Would have a substantial performance overhead. We'll see where it is going when we support star topologies and such.

patriknw replied:
It would actually be nice to configure this per replica, but here on the producer side we don't know which replica is the consumer (at least I couldn't see that)

patriknw replied:
yes, need to think this through again. we want to support a full mesh in cloud where indirect-replication=off, and edge connecting in to one or a few cloud replicas. edge consumers need indirect-replication=on

Maybe we could look at the allReplicas (or otherReplicas) and use that for deciding indirect. If it's included in the defined allReplicas then off, otherwise on. However, again, here on the producer side we don't know the replicaId of the consumer. Maybe we have to add that?

johanandren replied:
We want to indirectly replicate from cloud to edge, and from edge to other cloud replicas, so maybe it is more about knowing which replication goes/comes from where when deciding (I haven't completely thought this all the way through)?

patriknw commented on akka-projection-grpc/src/main/protobuf/akka/projection/grpc/event_producer.proto:
```diff
@@ -65,6 +65,10 @@ message InitReq {
+  // Used for Replicated Event Sourcing, the replica id of the consumer
+  string replica_id = 6;
+  // Used for Replicated Event Sourcing, other replicas that the consumer is connected to.
+  repeated string other_replica_ids = 7;
```
@johanandren I have tried this instead. These replica ids are included from the consumer in the InitReq. Then that is used on the producer side to decide if an event based on its origin should be emitted or handled as FilteredEvent.

patriknw replied:
For producer push it would be similar. These fields would be included in the ConsumerEventStart response from the consumer side and then used on the producer side.

johanandren replied:
Ok, seems reasonable.

patriknw commented on akka-projection-grpc/src/main/scala/akka/projection/grpc/replication/internal/EventOriginFilter.scala:
```diff
@@ -0,0 +1,41 @@
+        case Some(meta: ReplicatedEventMetadata) =>
+          !exclude(meta.originReplica)
+        case _ =>
+          // FIXME eventMetadata isn't loaded by backtracking
```
It's working aside from this little detail. For ordinary transformation, which we used for the origin filtering, we solve it by not applying transformation for backtracking and instead do that in loadEvent. It would be difficult to do the same for this, because then we would need the replica ids also in the load event request.

Do you think it's fine to change the r2dbc query to always load metadata, also for backtracking?

patriknw replied:
Sleeping on this. Feels fragile to not use the same origin filter from loadEvent also. Even though we only use loadEvent from backtracking events I think it should have the same transformation and filtering mechanism. I think it's best to include those additional replica id fields also in the loadEvent request.

> It would be difficult to do the same for this

no, that shouldn't be difficult

johanandren replied:
I agree, let's go with always passing metadata and applying the same transform/filter.

patriknw reviewed (commented):
New approach, would like feedback before continuing...

patriknw commented:
Incorporated the filter changes from https://github.com/akka/akka-projection/pull/1072 because there were too many changes in same place. Also added test for such filters.

johanandren commented:
Rebased and bumped Akka to get this ready for merge

patriknw commented:
Don't know what happened, but this didn't contain my latest work. Restored from history so now this should be ready.

patriknw commented on akka-projection-grpc/src/main/protobuf/akka/projection/grpc/event_consumer.proto:
```diff
@@ -59,6 +59,7 @@ message ConsumerEventInit {
 
 message ConsumerEventStart {
   repeated FilterCriteria filter = 1;
+  ReplicaInfo replica_info = 6;
```
added also to ConsumerEventStart but setting it and implementing corresponding filter on the producer side for RES must be done in @johanandren 's #1073 (or after that)

johanandren replied:
I'll do that in that PR once rebased on top of this after merged

johanandren reviewed (approved):
LGTM

patriknw commented:
Test fail was in IntegrationSpec, which I think is unrelated. https://github.com/akka/akka-projection/issues/905
However, I see it failing consistently locally, but not in main.

patriknw commented:
ah, that was because I changed log level. Some tests needed trace level. Should work now. 🤞
